### PR TITLE
Shipping fallback notes

### PR DESCRIPTION
**Pull request by masonmcelvain**
Closes https://github.com/iFixit/checkout-ui-extensions/issues/31

Implements the feature from shopify-v2 as described in https://github.com/iFixit/shopify-v2/issues/526

The translations should match [those from shopify-v2](https://github.com/iFixit/shopify-v2/blob/4e1c1341fb89d66c697f934b6b9b127e0be1ae47/src/locales/en.default.json#L606-L614).

## QA

Run local dev and preview `purchase.checkout.shipping-option-list.render-before`. If your cart meets any of the conditions outlined in https://github.com/iFixit/shopify-v2/issues/526, then you should see a warning banner with extra information like below. Please test all the conditions.

image

You might have to play with the [shipping rates](https://admin.shopify.com/store/ifixit-checkout-dev/settings/shipping) on the checkout dev store in order make it such that no shipping options show up. Alternatively, you could drop that check from the code with a diff like:

```diff
-   if (!isHeadlessStore(shop) || hasDeliveryOptions) {
+   if (!isHeadlessStore(shop)) {
      return null;
   }
```

You might also need to make certain products in stock, change their weights, or change their tags. That is all fine, it's a dev store.

### CODE REVIEW

**Comment by deltuh-vee** [comment]
dev_block ✊ 
[Variants](https://admin.shopify.com/store/ifixit-checkout-dev/apps/256-metafields-editor/productVariants/0?id=47582522507549&shop=ifixit-checkout-dev.myshopify.com) with metafield `us_only = 1` do not trigger the associated message.
CleanShot 2024-03-06 at 15 27 59@2x

**Comment by deltuh-vee** [comment]
@masonmcelvain Is there a way to get Mexico into the Country/Region dropdown?
CleanShot 2024-03-07 at 09 53 24

**Comment by deltuh-vee** [comment]
dev_block ✊ 
I'm not being stopped by shipping restrictions.
https://github.com/iFixit/checkout-ui-extensions/assets/5289715/785d623a-1b72-4727-8809-039e98283ee7

**Comment by masonmcelvain** [comment]
> dev_block I'm not being stopped by shipping restrictions.

Good observation. This feature intends to provide extra information for the customer though, it is not intended to prevent checkout progress. On our headless stores, these restrictions are built into our shipping rates. If a user's cart violates one of these restrictions, they will not be presented any shipping rates and therefore they won't be able to complete checkout. Sorry for not elaborating in the PR comment.

un_dev_block 👍🏻

**Comment by masonmcelvain** [comment]
Thanks for diving in Erin! Your steps all look correct to me, however these warnings will only show up if there are _no shipping options_ available. In your screenshot, there is 1 shipping method, "Standard International" for $5.99.

You can [edit the shipping rates](https://admin.shopify.com/store/ifixit-checkout-dev/settings/shipping) in Shopify to cause no shipping options to show up for certain regions if you want. I think @deltuh-vee was QAing that way.

If that's not working for you or is too time consuming, you could edit the code with a diff like this to get the warnings to show up even if there are shipping options.

`extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/ShippingFallbackNote.tsx`
```diff
-   if (!isHeadlessStore(shop) || hasDeliveryOptions) {
+   if (!isHeadlessStore(shop)) {
      return null;
   }
```

**Comment by masonmcelvain** [comment]
> I'm getting this single 401 in my console

I see this too sometimes. I think it is Shopify JS that is failing, and I don't think it is causing any problems for us.

**Comment by erinemay** [comment]
> these warnings will only show up if there are _no shipping options_ available. 

@deltuh-vee can you clarify how you were testing? I'm surprised to see shipping options available in the UI here: https://github.com/iFixit/checkout-ui-extensions/pull/59#issuecomment-1984215179. I don't think anything changed since that comment. 
image

**Comment by erinemay** [comment]
Thanks @masonmcelvain sorry about the back and forth. QA 👍 I confirmed all this based on Jeff's original spec:


- [x] Cart weight exceeds 150 pounds. Language String: Your cart exceeds our weight limit. Please contact iFixit Customer Support for a custom shipping solution.
- [x] Cart contains a Product with Tag BATTERY. Address is outside 48 states. Language String: Batteries may only be shipped within the contiguous USA at this time..
- [x] Cart contains a Product with Tag Dangerous. Address is outside 48 states. Language String: Batteries may only be shipped within the contiguous USA at this time.
- [x] Cart contains a Variant with metafield us_only = 1. Address country is not US. Language String: Due to shipping regulations, iFixit may only ship this product within the United States of America.
- [x] Cart contains a Variant with metafield does_not_ship_to_mexico= 1. Address country is Mexico. Language String: Due to shipping regulations, iFixit may not ship this product to Mexico..
- [x] Cart contains a Product with Tag BATTERY. Address is within the 48 states. Language String: Batteries may not be shipped via air freight or be delivered to PO Boxes, APOs, FPOs, or DPOs. If you are encountering this message in error, please contact iFixit Customer Support at support@ifixit.com.
- [x] None of these ☝️  - no warning.
- [x] Confirmed that the contiguous US warning shows for Alaska addresses. The us_only case that allows shipping to all US does not show up for Alaska. 

deploy_block 🌵 sorry if this info is somewhere and I didn't see it, but is there any sort of precedence or should we see all warnings at once? For example: If I trigger a BATTERY warning and the cart weight is over 150 lbs, I see only the weight warning.

**Comment by masonmcelvain** [comment]
> is there any sort of precedence

Good question. There _is_ supposed to be an order of precedence, but it's hard to tell since there is so much text in the original [issue](https://github.com/iFixit/shopify-v2/issues/526). Quoting it, Jeff said:

> It's likely that multiple conditions could be met, so I'll list the desired logic in order of importance

**Comment by masonmcelvain** [comment]
So cart weight should take precedence over anything else, then cart contains a battery, etc.

File: extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/ShippingFallbackNote.tsx [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+      key: 'does_not_ship_to_mexico',
+   });
+   const { cartWeightPounds, cartProductTags } = useCartExtras();
+   const deliveryGroup = useDeliveryGroupTarget();
+   const shippingAddress = useShippingAddress();
+   const shop = useShop();
+   const translate = useTranslate();
+
+   const hasDeliveryOptions =
+      deliveryGroup && deliveryGroup.deliveryOptions.length > 0;
```
[the remark this answers is not in the document]
Comment (masonmcelvain) [reply extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/ShippingFallbackNote.tsx:37]: You're right, but typescript complains about possibly comparing `undefined` to `0`.

**Comment by masonmcelvain** [comment]
> If I trigger a BATTERY warning and the cart weight is over 150 lbs, I see only the weight warning.

Sounds like the order of precedence is still being respected, so I'm going to un_deploy_block 👍🏻

**Pull request merged**